### DISCUSSION TITLE
Chatflow to use different memory for each LLM block

### DISCUSSION THREAD

**Discussion by rayzamgh** [langgenius/dify#4418] [Suggestion]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

I was trying out chatflows with multiple LLM blocks and was wondering why the processed data output seem to include the same memory for all LLM blocks?

Here is my chatflow configuration
![rhkjenrhkjre](https://github.com/langgenius/dify/assets/40740551/52a639b4-ec40-41f2-bae0-ed8bc67c1349)

Here is the fewshot sample for LLM block 1
![sample llm block 1](https://github.com/langgenius/dify/assets/40740551/386f74d0-0d2d-4451-850f-bd18074e325c)

Here is the processed data output from the first LLM block, notice that the fewshot example is correctly appended, but the memory section, particularly the output from the assistant is taken from the final answer block, instead of the output of LLM block 1
![srbkjrnslsrnh](https://github.com/langgenius/dify/assets/40740551/2da1cb44-90c7-4984-9846-4216a76ea70b)

### 2. Describe the feature you'd like to see

Currently i suspect the memory is shared accross all LLM blocks like so:

![currentconf drawio](https://github.com/langgenius/dify/assets/40740551/7850dbc0-70e4-4d6e-95e5-49396ec198d3)

Propose to separate them, such that each block has their own memory

![proposed drawio](https://github.com/langgenius/dify/assets/40740551/4bc207e7-0b48-4619-9b33-3019906f4593)

### 3. How will this feature improve your workflow or experience?

I think this will improve chatflow for a lot of people with more complex needs, particularly with multi agent requirement with separate function, in a single chatflow

### 5. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

**Comment by Garhorne0813** [comment]
I'm looking forward to that feature. Maybe I can use this feature to implement query rewriting for multiple rounds of conversations.

**Comment by wow-ww** [comment]
+1

**Comment by JianGuo1111** [comment]
I think the memory should only record the current LLM output, not all of it! All of the current LLM output is aggregated to the LLM that opens the memory.
